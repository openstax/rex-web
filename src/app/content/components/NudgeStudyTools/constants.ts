export const arrowDesktopHeight = 11.5;
export const arrowDesktopWidth = 15;
export const arrowMobileHeight = 6.2;
export const arrowMobileWidth = 5.2;
export const arrowRightMargin = 1;
export const arrowTopMargin = 1.6;

export const closeButtonSize = 4;
export const closeButtonDistanceFromContent = 1.6;
export const closeButtonMobileMargin = 3;

export const contentMarginTop = 1.6;
export const contentMarginLeft = 1;
export const contentWidth = 69;

export const mobileSpotlightPadding = .6;

export const cookieNudge = {
  // How many times nuge was shown
  counter: 'nudge_study_guides_counter',
  // When nudge was shown
  date: 'nudge_study_guides_date',
  // How many pages user has opened (opening the same page twice counts it as 2)
  pageCounter: 'nudge_study_guides_page_counter',
};

// Show nudge only for users which opened at least 2 pages
export const nudgeStudyToolsMinPageLimit = 2;

export const daysUntilCookieExpires = 365 * 20;

// Show nudge every 4 weeks: 4 * 7 days * 24 hours * 60 minutes * 60 seconds * 1000 ms
export const timeIntervalBetweenShowingNudgeInMs = 4 * 7 * 24 * 60 * 60 * 1000;
// Show nudge max 3 times
export const nudgeStudyToolsShowLimit = 3;
// Id of the element which should be highlighted
export const nudgeStudyToolsTargetId = 'nudge-study-tools';
export const mobileNudgeStudyToolsTargetId = 'mobile-menu-button';
