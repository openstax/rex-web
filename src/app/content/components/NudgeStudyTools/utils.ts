import { HTMLElement } from '@openstax/types/lib.dom';
import * as Cookies from 'js-cookie';
import React, { useState } from 'react';
import { useSelector } from 'react-redux';
import { useDebouncedWindowSize, useOnScrollTopOffset } from '../../../reactUtils';
import { assertDocument, remsToPx } from '../../../utils';
import { page as pageSelector } from '../../selectors';
import { hasStudyGuides } from '../../studyGuides/selectors';
import { Page } from '../../types';
import {
  arrowDesktopHeight,
  arrowRightMargin,
  arrowTopMargin,
  closeButtonDistanceFromContent,
  closeButtonSize,
  contentMarginLeft,
  contentMarginTop,
  contentWidth,
  cookieNudge,
  daysUntilCookieExpires,
  mobileNudgeStudyToolsTargetId,
  mobileSpotlightPadding,
  nudgeStudyToolsMinPageLimit,
  nudgeStudyToolsShowLimit,
  nudgeStudyToolsTargetId,
  timeIntervalBetweenShowingNudgeInMs,
} from './constants';

interface Positions {
  arrowLeft: number;
  arrowTopOffset: number;
  closeButtonLeft: number;
  closeButtonTopOffset: number;
  contentWrapperLeft: number;
  contentWrapperTopOffset: number;
  spotlightHeight: number;
  spotlightLeftOffset: number;
  spotlightTopOffset: number;
  spotlightWidth: number;
}

export const getPositions = (target: HTMLElement, isMobile: boolean): Positions => {
  const padding = remsToPx(mobileSpotlightPadding);
  const { top, left, height, width } = getBoundingRectOfNudgeTarget(target);
  const spotlightLeftOffset = left - (isMobile ? padding : 0);
  const spotlightWidth = width + (isMobile ? padding * 2 : 0);
  const arrowTopOffset = top + height + remsToPx(arrowTopMargin);
  // left edge of arrow image should be on the middle of the spotlight (adjusted for a margin)
  const arrowLeft = spotlightLeftOffset + (spotlightWidth / 2) - remsToPx(arrowRightMargin);

  const contentWrapperTopOffset = arrowTopOffset
    + remsToPx(arrowDesktopHeight)
    + remsToPx(contentMarginTop);
  const contentWrapperLeft = spotlightLeftOffset + spotlightWidth + remsToPx(contentMarginLeft);
  const closeButtonLeft = contentWrapperLeft + remsToPx(contentWidth) + remsToPx(closeButtonDistanceFromContent);
  const closeButtonTopOffset = contentWrapperTopOffset
    - remsToPx(closeButtonSize)
    - remsToPx(closeButtonDistanceFromContent);

  return {
    arrowLeft,
    arrowTopOffset,
    closeButtonLeft,
    closeButtonTopOffset,
    contentWrapperLeft,
    contentWrapperTopOffset,
    spotlightHeight: height,
    spotlightLeftOffset,
    spotlightTopOffset: top,
    spotlightWidth,
  };
};

// Target may have display set to `contents` so we are calculatings rect for children of passed target.
const getBoundingRectOfNudgeTarget = (target: HTMLElement) => {
  const { top, bottom, left, right } = Array.from(target.children).reduce((rects, child) => {
    const rect = child.getBoundingClientRect();
    rects.top = Math.min(rects.top, rect.top);
    rects.bottom = Math.max(rects.bottom, rect.bottom);
    rects.left = Math.min(rects.left, rect.left);
    rects.right = Math.max(rects.right, rect.right);
    return rects;
    // Starting values depends on if we checks for Math.max or Math.min
  }, { top: 9999, bottom: 0, left: 9999, right: 0 });

  return {
    height: bottom - top,
    left,
    right,
    top,
    width: right - left,
  };
};

export const usePositions = (isMobile: boolean) => {
  const [windowWidth] = useDebouncedWindowSize();
  const scrollTopOffset = useOnScrollTopOffset();
  const [positions, setPositions] = React.useState<Positions | null>(null);
  const studyGuides = useSelector(hasStudyGuides);

  React.useEffect(() => {
    const document = assertDocument();
    const target = document.querySelector(
      `#${isMobile ? mobileNudgeStudyToolsTargetId : nudgeStudyToolsTargetId}`
    ) as HTMLElement | null;
    if (target) {
      // Make sure that we calculate positions with body overflow set to hidden
      // because it causes scrollbar to hide which results in different positions.
      const prevOverflow = document.body.style.overflow;
      document.body.style.overflow = 'hidden';
      setPositions(getPositions(target, isMobile));

      // Resets to the value from before calculations. We want this style change to be handled
      // directly in the component.
      document.body.style.overflow = prevOverflow;
    }
    return () => setPositions(null);
  }, [scrollTopOffset, studyGuides, windowWidth, isMobile]);

  return positions;
};

export const getCounterCookie = () => {
  return Number(Cookies.get(cookieNudge.counter) || 0);
};

export const getDateCookie = () => {
  const lastShownDate = Cookies.get(cookieNudge.date);
  return lastShownDate ? new Date(lastShownDate) : undefined;
};

export const passedTimeInterval = () => {
  const lastShownDate = getDateCookie();
  return !lastShownDate
    ? true
    : (Date.now() - lastShownDate.getTime()) > timeIntervalBetweenShowingNudgeInMs
  ;
};

export const getPageCounterCookie = () => {
  return Number(Cookies.get(cookieNudge.pageCounter) || 0);
};

export const incrementPageCounterCookie = () => {
  const counter = getPageCounterCookie();
  const newValue = counter + 1;
  Cookies.set(cookieNudge.pageCounter, newValue.toString(), {expires: daysUntilCookieExpires});
  return newValue;
};

export const shouldDisplayNudgeStudyTools = (): boolean => {
  const counter = getCounterCookie();
  const numberOfPagesOpenedByUser = getPageCounterCookie();

  return counter < nudgeStudyToolsShowLimit
    && passedTimeInterval()
    && (numberOfPagesOpenedByUser >= nudgeStudyToolsMinPageLimit);
};

// Set required cookies and reset opened page counter
export const setNudgeStudyToolsCookies = () => {
  const now = new Date();
  const counter = getCounterCookie();
  Cookies.set(cookieNudge.counter, (counter + 1).toString(), {expires: daysUntilCookieExpires});
  Cookies.set(cookieNudge.date, now.toString(), {expires: daysUntilCookieExpires});
  Cookies.remove(cookieNudge.pageCounter);
};

export const useIncrementPageCounter = () => {
  const page = useSelector(pageSelector);
  const [counter, updateCounter] = useState<number>(getPageCounterCookie());
  const pageRef = React.useRef<Page>();

  React.useEffect(() => {
    if (page && page !== pageRef.current && counter <= nudgeStudyToolsMinPageLimit) {
      pageRef.current = page;
      updateCounter(incrementPageCounterCookie());
    }
  }, [page, updateCounter, counter]);

  return counter;
};
