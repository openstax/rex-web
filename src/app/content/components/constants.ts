import theme from '../../theme';
import { BookWithOSWebData } from '../types';

export const bookBannerDesktopBigHeight = 13;
export const bookBannerDesktopMiniHeight = 7;
export const bookBannerMobileBigHeight = 10.4;
export const bookBannerMobileMiniHeight = 6;

export const scrollOffset = 3;

export const sidebarDesktopWidth = 37.5;
export const sidebarMobileWidth = 30.4;
export const sidebarTransitionTime = 300;

export const searchResultsBarDesktopWidth = 37.5;
export const searchResultsBarMobileWidth = 33.8;

export const toolbarIconColor = theme.color.primary.gray;

export const mobileSearchContainerMargin = 1;
export const toolbarWidth = 8;
export const toolbarSearchInputHeight = 3.2;
export const toolbarSearchInputMobileHeight = 3.2;
export const toolbarSearchInputMobileWidth = 18.5;
export const toolbarToggleSearchBarTextMobileWidth = 11.2;
export const toolbarButtonMargin = 2;
export const toolbarHrHeight = 0.1;
export const sidebarDesktopWithToolbarWidth = searchResultsBarDesktopWidth + toolbarWidth;
export const toolbarMobileSearchWrapperHeight = toolbarSearchInputMobileHeight
  + (mobileSearchContainerMargin * 2)
  + toolbarHrHeight;

export const topbarDesktopHeight = 5;
export const topbarMobileHeight = 4;
export const toolbarMobileExpandedHeight = topbarMobileHeight
  + toolbarMobileSearchWrapperHeight;

export const verticalNavbarMaxWidth = 8;

export const searchSidebarTopOffset = bookBannerMobileMiniHeight
  + toolbarMobileExpandedHeight;

export const contentTextWidth = 82.5;

export const mainContentBackground = '#fff';

export const maxContentGutter = 6;
export const contentWrapperMaxWidth = contentTextWidth + sidebarDesktopWidth + verticalNavbarMaxWidth;

export const defaultTheme = 'blue' as BookWithOSWebData['theme'];
