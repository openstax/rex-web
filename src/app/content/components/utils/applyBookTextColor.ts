import { css } from 'styled-components/macro';
import theme from '../../../theme';
import { BookWithOSWebData } from '../../types';

export const applyBookTextColor = (props: {colorSchema: BookWithOSWebData['theme'] | undefined }) =>
  props.colorSchema && css`
    color: ${theme.color.primary[props.colorSchema].foreground};
  `;
