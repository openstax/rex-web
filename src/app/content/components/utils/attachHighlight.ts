import Highlighter, { Highlight, SerializedHighlight } from '@openstax/highlighter';
import Sentry, { Severity } from '../../../../helpers/Sentry';

const attachHighlight = <T extends Highlight | SerializedHighlight>(
  highlight: T,
  highlighter: Highlighter,
  errorMsg?: (highlight: T) => string
): Highlight | undefined => {
  let attachedHighlight: Highlight | undefined;

  try {
    highlighter.highlight(highlight);
    attachedHighlight = highlighter.getHighlight(highlight.id);
  } catch (e) {
    Sentry.captureException(e);
  }

  if (!attachedHighlight || !attachedHighlight.isAttached()) {
    Sentry.captureException(
      new Error(errorMsg ?  errorMsg(highlight) : `Highlight with id: ${highlight.id} has not been attached.`),
      Severity.Warning
    );
    attachedHighlight = undefined;
  }

  return attachedHighlight;
};

export default attachHighlight;
