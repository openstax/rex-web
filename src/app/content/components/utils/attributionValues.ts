import { Book } from '../../types';

// date is initialized as UTC, conversion to local time can change the date.
// this compensates
export function compensateForUTC(date: Date): void {
  date.setMinutes(date.getMinutes() + date.getTimezoneOffset());
}

export function getPublishDate(book: Book): Date | null {
  if (!('publish_date' in book)) {
    return null;
  }
  const date = new Date(book.publish_date);

  compensateForUTC(date);
  return date;
}

export function getAuthors(book: Book) {
  if (!('authors' in book)) {
    return [];
  }
  const seniorAuthors = book.authors.filter(
    author => author.value.senior_author
  );

  return seniorAuthors.length > 0 ? seniorAuthors : book.authors.slice(0, 2);
}

export const bookIdsWithSpecialAttributionText: {
  [key: string]: {
    copyrightHolder?: string;
    originalMaterialLink?: null | string;
  };
} = {
  '1b4ee0ce-ee89-44fa-a5e7-a0db9f0c94b1': {
    copyrightHolder: 'The Michelson 20MM Foundation',
  },
  '394a1101-fd8f-4875-84fa-55f15b06ba66': {
    copyrightHolder: 'Texas Education Agency (TEA)',
    originalMaterialLink: 'https://www.texasgateway.org/book/tea-statistics',
  },
  'cce64fde-f448-43b8-ae88-27705cceb0da': {
    copyrightHolder: 'Texas Education Agency (TEA)',
    originalMaterialLink: 'https://www.texasgateway.org/book/tea-physics',
  },
};

export function attributionValues(book: Book) {
  const bookPublishDate = getPublishDate(book);
  const authorsToDisplay = getAuthors(book);
  const locale = book.language;

  return {
    bookTitle: book.title,
    publisher: locale === 'pl' ? 'OpenStax Poland' : 'OpenStax',
    language: new Intl.DisplayNames([locale], { type: 'language' }).of(locale) as string,
    bookPublishDate,
    authorsToDisplay,
  };
}
