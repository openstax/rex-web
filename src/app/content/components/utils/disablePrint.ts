import { css } from 'styled-components';

export const disablePrint = css`@media print { display: none; }`;
