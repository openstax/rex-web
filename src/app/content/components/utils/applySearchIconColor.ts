import { css } from 'styled-components/macro';
import theme from '../../../theme';
import { toolbarIconColor } from '../constants';

const primaryColors = theme.color.primary;

export const applySearchIconColor = (colorSchema: keyof typeof primaryColors | null) =>
  colorSchema ? css`
    color: ${primaryColors[colorSchema].foreground};

    :hover,
    :focus {
        color: ${primaryColors[colorSchema].foreground};
        transition: color 200ms;

        ::before {
            content:"";
            display: block;
            height: 100%;
            position: absolute;
            top: 0;
            right: 0;
            min-width: 45px;
            background-color: rgba(0, 0, 0, 0.06)
            ;
        }
    }
  ` : css`
  color: ${toolbarIconColor.base};
`;
