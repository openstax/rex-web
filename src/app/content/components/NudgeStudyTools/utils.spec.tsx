import { DOMRect, HTMLElement } from '@openstax/types/lib.dom';
import * as Cookies from 'js-cookie';
import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import createTestStore from '../../../../test/createTestStore';
import { shortPage } from '../../../../test/mocks/archiveLoader';
import { Store } from '../../../types';
import { assertDocument, assertWindow } from '../../../utils';
import { receivePage } from '../../actions';
import * as studyGuidesSelect from '../../studyGuides/selectors';
import * as constants from './constants';
import * as utils from './utils';

describe('usePositions', () => {
  let target: HTMLElement;
  let store: Store;
  // tslint:disable-next-line: variable-name
  let Component: (props: { isMobile: boolean }) => JSX.Element;
  const mockRect = { bottom: 228, height: 25, left: 951, right: 1233, top: 203, width: 282 } as any as DOMRect;

  beforeEach(() => {
    const document = assertDocument();
    target = document.createElement('div');
    target.setAttribute('id', constants.nudgeStudyToolsTargetId);
    const child = document.createElement('div');
    target.append(child);
    jest.spyOn(child, 'getBoundingClientRect')
      .mockReturnValue(mockRect);
    document.body.appendChild(target);

    store = createTestStore();
    Component = (props) => {
      const positions = utils.usePositions(props.isMobile);
      return <React.Fragment>
        {JSON.stringify(positions)}
      </React.Fragment>;
    };
  });

  afterEach(() => {
    target.remove();
  });

  it('return null if target was not found', () => {
    target.remove();

    const component = renderer.create(<Provider store={store}>
      <Component isMobile={false} />
    </Provider>);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('return positions and updates body overflow to previous value after finishing calculations', () => {
    jest.spyOn(studyGuidesSelect, 'hasStudyGuides')
      .mockReturnValue(true);

    Object.defineProperty(assertWindow(), 'innerWidth', {
      value: 1900,
    });

    const component = renderer.create(<Provider store={store}>
      <Component isMobile={false} />
    </Provider>);

    renderer.act(() => {
      assertDocument().body.style.overflow = 'scroll';
    });

    component.update(<Provider store={store}>
      <Component isMobile={false} />
    </Provider>);

    expect(component.toJSON()).toMatchSnapshot();

    // Coverage for useEffect return function
    component.unmount();

    renderer.act(() => {
      assertDocument().body.style.overflow = 'scroll';
    });
  });

  it('returns different positions depends on isMobile and windowWidth', () => {
    // Default values
    expect(utils.getPositions(target, false, 1900)).toEqual({
      arrowLeft: 952,
      arrowTopOffset: 248,
      closeButtonLeft: 1253,
      closeButtonTopOffset: 348,
      contentWrapperRight: 657,
      contentWrapperTopOffset: 388,
      spotlightHeight: 45,
      spotlightLeftOffset: 941,
      spotlightTopOffset: 193,
      spotlightWidth: 302,
    });

    // Change of windowWidth affects only contentWrapperRight
    expect(utils.getPositions(target, false, 1200)).toEqual({
      arrowLeft: 952,
      arrowTopOffset: 248,
      closeButtonLeft: 1253,
      closeButtonTopOffset: 348,
      // this is less by 700px (1900 - 1200 = 700)
      contentWrapperRight: -43,
      contentWrapperTopOffset: 388,
      spotlightHeight: 45,
      spotlightLeftOffset: 941,
      spotlightTopOffset: 193,
      spotlightWidth: 302,
    });

    // Values when isMobile prop is passed (since we are not changing windowWidth only some of the values will change)
    expect(utils.getPositions(target, true, 1900)).toEqual({
      arrowLeft: 1050,
      arrowTopOffset: 248,
      closeButtonLeft: 1253,
      // Close button adjusted to contentWrapperTopOffset
      closeButtonTopOffset: 280,
      contentWrapperRight: 657,
      // In mobile arrow is smaller so content is closer to the top
      contentWrapperTopOffset: 320,
      spotlightHeight: 45,
      spotlightLeftOffset: 941,
      spotlightTopOffset: 193,
      spotlightWidth: 302,
    });
  });
});

describe('cookies helpers', () => {
  beforeEach(() => {
    Cookies.remove(constants.cookieNudge.date);
    Cookies.remove(constants.cookieNudge.counter);
    Cookies.remove(constants.cookieNudge.pageCounter);
  });

  it('getCounterCookie gets counter or 0', () => {
    expect(utils.getCounterCookie()).toEqual(0);
    Cookies.set(constants.cookieNudge.counter, '2');
    expect(utils.getCounterCookie()).toEqual(2);
  });

  it('getDateCookie gets date or undefined', () => {
    expect(utils.getDateCookie()).toEqual(undefined);
    const date = new Date();
    Cookies.set(constants.cookieNudge.date, date.toString());
    // date.toString() will remove ms from the date and jest's toEqual will find that as a difference
    // so we comapre with new Date(date.toString())
    expect(utils.getDateCookie()).toEqual(new Date(date.toString()));
  });

  it('getPageCounterCookie gets counter or undefined', () => {
    expect(utils.getPageCounterCookie()).toEqual(0);
    Cookies.set(constants.cookieNudge.pageCounter, '3');
    expect(utils.getPageCounterCookie()).toEqual(3);
  });

  it('passedTimeInterval - return false if did not pass time interval', () => {
    expect(utils.passedTimeInterval()).toEqual(true);
    Cookies.set(constants.cookieNudge.date, new Date().toString());
    expect(utils.passedTimeInterval()).toEqual(false);
  });

  it('passedTimeInterval - return true if passed allowed time interval or last date is undefined', () => {
    expect(utils.passedTimeInterval()).toEqual(true);
    const now = new Date();
    const passed = new Date(now.getTime() - constants.timeIntervalBetweenShowingNudgeInMs - 10000);
    Cookies.set(constants.cookieNudge.date, passed.toString());
    expect(utils.passedTimeInterval()).toEqual(true);
  });

  it('incrementPageCounterCookie increments page counter cookie', () => {
    expect(utils.getPageCounterCookie()).toEqual(0);
    utils.incrementPageCounterCookie();
    utils.incrementPageCounterCookie();
    expect(utils.getPageCounterCookie()).toEqual(2);
  });

  it('shouldDisplayNudgeStudyTools - return true only if all criteria passes', () => {
    expect(utils.shouldDisplayNudgeStudyTools()).toEqual(false);

    const now = new Date();
    const passed = new Date(now.getTime() - constants.timeIntervalBetweenShowingNudgeInMs - 10000);
    Cookies.set(constants.cookieNudge.date, passed.toString());
    expect(utils.shouldDisplayNudgeStudyTools()).toEqual(false);

    Cookies.set(constants.cookieNudge.pageCounter, constants.nudgeStudyToolsMinPageLimit.toString());
    expect(utils.shouldDisplayNudgeStudyTools()).toEqual(true);

    Cookies.set(constants.cookieNudge.counter, constants.nudgeStudyToolsShowLimit.toString());
    expect(utils.shouldDisplayNudgeStudyTools()).toEqual(false);
  });

  it('setNudgeStudyToolsCookies', () => {
    Cookies.set(constants.cookieNudge.pageCounter, '1');
    const counter = utils.getCounterCookie();
    const pageCounter = utils.getPageCounterCookie();
    const date = utils.getDateCookie();
    expect(counter).toEqual(0);
    expect(pageCounter).toEqual(1);
    expect(date).toEqual(undefined);

    utils.setNudgeStudyToolsCookies();

    expect(utils.getCounterCookie()).toEqual(1);
    expect(utils.getPageCounterCookie()).toEqual(0);
    expect(utils.getDateCookie()).toBeTruthy();
  });

  it('useIncrementPageCounter - increments page counter cookie to until it reaches limit', () => {
    const store = createTestStore();
    // tslint:disable-next-line: variable-name
    const Component = () => {
      utils.useIncrementPageCounter();
      return <div/>;
    };

    renderer.create(<Provider store={store}>
      <Component />
    </Provider>);

    expect(utils.getPageCounterCookie()).toEqual(0);
    renderer.act(() => {
      store.dispatch(receivePage({...shortPage, references: []}));
    });

    expect(utils.getPageCounterCookie()).toEqual(1);
    renderer.act(() => {
      store.dispatch(receivePage({...shortPage, references: []}));
    });

    expect(utils.getPageCounterCookie()).toEqual(2);
    renderer.act(() => {
      store.dispatch(receivePage({...shortPage, references: []}));
    });

    expect(utils.getPageCounterCookie()).toEqual(3);
  });
});
