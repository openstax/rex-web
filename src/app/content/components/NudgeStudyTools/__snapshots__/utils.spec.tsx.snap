// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`usePositions return null if target was not found 1`] = `"null"`;

exports[`usePositions return positions and updates body overflow to previous value after finishing calculations 1`] = `"{\\"arrowLeft\\":1082,\\"arrowTopOffset\\":244,\\"closeButtonLeft\\":1949,\\"closeButtonTopOffset\\":319,\\"contentWrapperLeft\\":1243,\\"contentWrapperTopOffset\\":375,\\"spotlightHeight\\":25,\\"spotlightLeftOffset\\":951,\\"spotlightTopOffset\\":203,\\"spotlightWidth\\":282}"`;

exports[`usePositions returns different positions depending on isMobile: 
    Object {
      "arrowLeft": 1082,
      "arrowTopOffset": 244,
      "closeButtonLeft": 1949,
      "closeButtonTopOffset": 319,
      "contentWrapperLeft": 1243,
      "contentWrapperTopOffset": 375,
      "spotlightHeight": 25,
      "spotlightLeftOffset": 951,
      "spotlightTopOffset": 203,
      "spotlightWidth": 282,
    }
   1`] = `
Object {
  "arrowLeft": 1082,
  "arrowTopOffset": 244,
  "closeButtonLeft": 1949,
  "closeButtonTopOffset": 319,
  "contentWrapperLeft": 1243,
  "contentWrapperTopOffset": 375,
  "spotlightHeight": 25,
  "spotlightLeftOffset": 951,
  "spotlightTopOffset": 203,
  "spotlightWidth": 282,
}
`;

exports[`usePositions selects hamburger menu as nudge study tools target if isMobile prop is passed 1`] = `"null"`;
